core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/StreamManagingStage.java

```diff
@@ -43,16 +47,44 @@ public StreamManagingStage(RequestPipeline<SdkHttpFullRequest, Response<OutputT>
 
     @Override
     public Response<OutputT> execute(SdkHttpFullRequest request, RequestExecutionContext context) throws Exception {
+        ClosingStreamProvider toBeClosed = null;
         if (request.contentStreamProvider().isPresent()) {
-            request = request.toBuilder()
-                             .contentStreamProvider(createManagedStream(request.contentStreamProvider().get())).build();
+            toBeClosed = createManagedProvider(request.contentStreamProvider().get());
+            request = request.toBuilder().contentStreamProvider(toBeClosed).build();
+        }
+        try {
+            InterruptMonitor.checkInterrupted();
+            return wrapped.execute(request, context);
+        } finally {
+            if (toBeClosed != null) {
+                toBeClosed.closeCurrentStream();
```

Comment:
Does this close inputstream provided by customer i.e. `RequestBody.fromInputStream` or `RequestBody.fromContentProvider`?

Reply:
> RequestBody.fromInputStream

No, the method disables close on the stream: https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/RequestBody.java#L119

> RequestBody.fromContentProvider

This depends. `StreaManagingStage` always calls `close()` on the last stream it got from the `ContentStreamProvider`, so unless the `close()` is disabled on the stream, then it will get closed.

Thoughts? Should the provider be wrapped so that it wraps stream from `newStream()` and disables close?

Reply:
I'm sort of conflicted on it since the interface says nothing about closing the *last* stream https://github.com/aws/aws-sdk-java-v2/blob/master/http-client-spi/src/main/java/software/amazon/awssdk/http/ContentStreamProvider.java#L31

Reply:
Interface says `may close or otherwise discard the stream returned from the previous invocation`. So I am okay with closing the stream if customer didn't explicitly tell us to disable close.